grpc-dump/dump/recorded_server_stream.go

```diff
@@ -1,17 +1,43 @@
 package dump
 
 import (
-	"github.com/bradleyjkemp/grpc-tools/internal"
-	"google.golang.org/grpc"
 	"sync"
 	"time"
+
+	"github.com/bradleyjkemp/grpc-tools/internal"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
 )
 
 // recordedServerStream wraps a grpc.ServerStream and allows the dump interceptor to record all sent/received messages
 type recordedServerStream struct {
 	sync.Mutex
 	grpc.ServerStream
-	events []*internal.Message
+	events   []*internal.Message
+	headers  metadata.MD
+	trailers metadata.MD
+}
+
+func (ss *recordedServerStream) SendHeader(headers metadata.MD) error {
```

Comment:
Unrelated to this PR but I'd be interested to get your thoughts on https://github.com/bradleyjkemp/grpc-tools/issues/6 which would effectively be replacing this batching of events with logging events as they happen.

Benefits would be that the dump would much more accurately represent what happened during an RPC (especially streaming ones) but comes at a cost of being a bit harder to reassembly the output

The result would look a bit like https://github.com/bradleyjkemp/grpc-tools/blob/21195c2497c51fab93b0adcac105a6da83f1ccb3/integration_test/.snapshots/TestIntegration.json. Would this negatively impact usability for you?

Reply:
Great idea. This would not really change the way I am working with `grpc-tools` right now, but anyway would be a great improvement. Of course we would need to tie the response headers / trailers with the `id` as well.